### PR TITLE
SI-6384 - avoid crash due to optimization in 2.10 CPS plugin

### PR DESCRIPTION
**Pull request by phaller**
This fixes a regression in the 2.9.x branch (code combining CPS and try-catch crashing the compiler). The fix is simply undoing an optimization that was done in the 2.10 CPS plugin (replacing a function with only a match by a match on an empty tree; this optimization seems to work only with virtpatmat).

Review by @adriaanm

### CODE REVIEW

**Comment by phaller** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
Rangepos tests always fail on 2.9.x branch. That one of things that are broken with our current bot.

**Comment by phaller** [comment]
Good to know. Thanks for pointing it out, Greg!

**Comment by jsuereth** [comment]
So this is passing.  Any reviewers?

**Comment by adriaanm** [comment]
Sorry this slipped past. Will review today.

File: src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala [hunk 1]
```diff
@@ -191,14 +191,19 @@ abstract class SelectiveCPSTransform extends PluginComponent with
 
           val pos = catches.head.pos
           val funSym = currentOwner.newValueParameter(pos, cpsNames.catches).setInfo(appliedType(PartialFunctionClass.tpe, List(ThrowableClass.tpe, targettp)))
+          val argSym = currentOwner.newValueParameter(pos, cpsNames.ex).setInfo(ThrowableClass.tpe)
```
Comment (adriaanm) [inline src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala]: I think the argument symbol should be owned by the function's symbol.

File: src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala [hunk 2]
```diff
@@ -191,14 +191,19 @@ abstract class SelectiveCPSTransform extends PluginComponent with
...
+          val fun = Function(List(ValDef(argSym)), rhs)
           val funDef = localTyper.typed(atPos(pos) {
-            ValDef(funSym, Match(EmptyTree, catches1))
+            ValDef(funSym, fun)
           })
           val expr2 = localTyper.typed(atPos(pos) {
             Apply(Select(expr1, expr1.tpe.member(cpsNames.flatMapCatch)), List(Ident(funSym)))
           })
 
           val exSym = currentOwner.newValueParameter(pos, cpsNames.ex).setInfo(ThrowableClass.tpe)
```
Comment (adriaanm) [inline src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala]: replace `currentOwner` by `fun.symbol` and drop the next line

**Comment by phaller** [comment]
@adriaanm: can you have another look?

**Comment by phaller** [comment]
The problem is that before I have a reference to `fun` I've already used `argSym` twice:

``` scala
val argSym = currentOwner.newValueParameter(pos, cpsNames.ex).setInfo(ThrowableClass.tpe)
val rhs = Match(Ident(argSym), catches1)
val fun = Function(List(ValDef(argSym)), rhs)
```

It'd be great if we could avoid setting the owner afterwards, but I'm not sure how we could do that. Maybe by creating a synthetic `TermName` to be used in both the `Ident` and the `ValDef`?

**Comment by phaller** [comment]
I found a solution which avoids setting the owner explicitly (see latest commit). I agree this looks much cleaner.

**Comment by phaller** [comment]
@adriaanm: is the latest revision good to go?

**Comment by adriaanm** [comment]
Great! Thanks for pushing it all the way. LGTM

**Comment by adriaanm** [comment]
PLS REBUILD pr-rangepos

(it doesn't look like a spurious failure, but trying anyway)

**Comment by gkossakowski** [comment]
`rangepos` always fails on 2.9.x branch because jenkins/pr-validator are not set up properly. That's one of many things we need to fix in our infrastructure.

**Comment by adriaanm** [comment]
ah, sorry I didn't know that

merging now then

**Pull request merged**